src/downloadAzureProject/setupProjectFolder.ts

```diff
@@ -0,0 +1,101 @@
+import { getNameFromId } from '../utils/azure';
+import { requestUtils } from '../utils/requestUtils';
+import { getRequiredQueryParameter, HandleUriActions } from './handleUri';
+
+export async function setupProjectFolder(uri: vscode.Uri, vsCodeFilePathUri: vscode.Uri, context: IActionContext): Promise<void> {
+    const parsedQuery: querystring.ParsedUrlQuery = querystring.parse(uri.query);
+    const resourceId: string = getRequiredQueryParameter(parsedQuery, 'resourceId');
+    const devContainerName: string = getRequiredQueryParameter(parsedQuery, 'devcontainer');
+    const language: string = getRequiredQueryParameter(parsedQuery, 'language');
+    const action: string = getRequiredQueryParameter(parsedQuery, 'action');
+    const downloadAppContent: boolean = action === HandleUriActions.downloadContentAndSetupProject;
+
+    const toBeDeletedFolderPathUri: vscode.Uri = vscode.Uri.joinPath(vsCodeFilePathUri, 'temp');
+
+    try {
+        const functionAppName: string = getNameFromId(resourceId);
+        const downloadFilePath: string = vscode.Uri.joinPath(toBeDeletedFolderPathUri, `${functionAppName}.zip`).fsPath;
+
+        await vscode.window.withProgress({ location: vscode.ProgressLocation.Notification, title: localize('settingUpFunctionAppLocalProjInfoMessage', `Setting up project for function app '${functionAppName}' with language '${language}'.`) }, async () => {
+            if (downloadAppContent) {
+                // NOTE: We don't want to download app content for compiled languages.
+                const slotTreeItem: SlotTreeItemBase | undefined = await ext.tree.findTreeItem(resourceId, { ...context, loadAll: true });
+                const hostKeys: WebSiteManagementModels.HostKeys | undefined = await slotTreeItem?.client.listHostKeys();
+                const defaultHostName: string | undefined = slotTreeItem?.client.defaultHostName;
+
+                if (!!hostKeys && hostKeys.masterKey && defaultHostName) {
+                    const requestOptions: RequestPrepareOptions = {
+                        method: 'GET',
+                        url: `https://${defaultHostName}/admin/functions/download?includeCsproj=true&includeAppSettings=true`,
+                        headers: { 'x-functions-key':  hostKeys.masterKey }
+                    };
+                    await requestUtils.downloadFile(requestOptions, downloadFilePath);
+                }
+            }
+
+            const projectFilePathUri: vscode.Uri = vscode.Uri.joinPath(vsCodeFilePathUri, `${functionAppName}`);
+            const projectFilePath: string = projectFilePathUri.fsPath;
+            const devContainerFolderPathUri: vscode.Uri = vscode.Uri.joinPath(projectFilePathUri, '.devcontainer');
+            if (downloadAppContent) {
+                // tslint:disable-next-line: no-unsafe-any
+                await extract(downloadFilePath, { dir: projectFilePath });
+            }
+            await requestUtils.downloadFile(
```

Comment:
Sure you can set up a meeting to discuss the dev container/docker stuff. 

> When customers create a compiled language functionApp from Azure Portal, we don't allow then to create or manage any functions from the Portal and point them to using VSCode instead. This will help those customers to setup their local projects with all necessary packages and devcontainer and they can easily add functions and then publish to azure from VSCode. We will have all this documented on our end on the Azure Portal as well as any documentations needed.

Other than the reasons I've mentioned so far, another reason to avoid dev containers is to keep portal docs consistent with our [getting started tutorial](https://docs.microsoft.com/azure/azure-functions/create-first-function-vs-code-csharp) which has never used them.

Reply:
I have setup a meeting for 1/25 to discuss these open docker and dev container related comments.

Reply:
Discussed on the meeting